core/unix/signal.c

```diff
@@ -4072,11 +4072,15 @@ record_pending_signal(dcontext_t *dcontext, int sig, kernel_ucontext_t *ucxt,
          * pending or delete an old b/c we might mess up the state so we
          * just drop this one: should only happen for alarm signal
          */
-        (info->accessing_sigpending && !info->nested_pending_ok &&
+        (((info->accessing_sigpending && !info->nested_pending_ok) ||
+          /* It's risky to deliver to an exiting thread: data structs are in unstable
+           * states (i#4438).  Just drop if we can.
+           */
+          dcontext->is_exiting) &&
          /* we do want to report a crash in receive_pending_signal() */
          (can_always_delay[sig] ||
```

Comment:
Is delaying the signal different than dropping?

Reply:
The idea is that a signal that can be delayed is asynchronous and so *could* have arrived after the thread exited...so dropping it seems reasonable.  That's why the two terms are linked.